Added Ctrl+Click support for opening context menu on mac

HermanKirshin opened the pull request:
## What does the pull request do?
Fixes #12828 

I recreated previous request (see #12829), commit was reworked according to discussions, now it has only minimal  sufficient changes.
I removed handling PointerPressed event that was performed in previous PR. So it will not cover some cases like having context menu on control that has some logic on left pointerpressed, but these are rare cases that can be considered separately.

## What is the current behavior?
Ctrl+Click on MacOS does nothing

## What is the updated/expected behavior with this PR?
Ctrl+Click now opens context menu on MacOS

## Fixed issues
Fixes #12828

herman-kirshin committed:
Added Ctrl+Click support for opening context menu on mac

cla-avalonia commented:
@herman-kirshin,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

HermanKirshin commented:
@cla-avalonia agree

kekekeks commented:
1) WPF's MouseGesture's [MouseAction](https://learn.microsoft.com/en-us/dotnet/api/system.windows.input.mouseaction?view=windowsdesktop-8.0) is more sophisticated
2) the PointerGesture naming is rather confusing, since we have gesture recognition elsewhere. Maybe we should have PointerAction and PointerActionGesture? Not quite sure about the naming, need to discuss the API design more.

maxkatz6 commented on src/Avalonia.Base/Input/PointerGesture.cs:
```diff
@@ -0,0 +1,103 @@
+        return !Equals(left, right);
+    }
+
+    public MouseButton Button { get; }
```
It should be a 
```
public PointerUpdateKind PointerAction { get; }
```

maxkatz6 commented on src/Avalonia.Base/Input/PointerGesture.cs:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    public static bool operator ==(PointerGesture? left, PointerGesture? right)
```
This class can be a simple record from how it looks.

HermanKirshin replied:
done

kekekeks commented:
Also, there is a point about the entire thing being a gesture recognizer.

maxkatz6 commented on src/Avalonia.Base/Input/PointerGesture.cs:
```diff
@@ -0,0 +1,103 @@
+
+        if (KeyModifiers.HasAllFlags(KeyModifiers.Control))
+        {
+            s.Append("Ctrl");
```
We recently accepted a new key formatter API that helps to avoid hardcoding these key strings. It should be integrated here as well https://github.com/AvaloniaUI/Avalonia/pull/15828

HermanKirshin replied:
Fixed - in latest commits I just reuse the code from KeyGesture.
But now I have some doubts about string format. Mouse buttons have rather short names, so
Ctrl+Left may look confusing, especially because there may be the same KeyGesture. Need to either remove it for now, or agree on some distinguishable button naming.

kekekeks commented:
Also, how was the problem with Ctrl+Click being used for list item selection resolved?

herman-kirshin committed:
code review fixes for PlatformPointerConfiguration

herman-kirshin committed:
Added click count to PointerActionGesture

HermanKirshin commented:
> 1. WPF's MouseGesture's [MouseAction](https://learn.microsoft.com/en-us/dotnet/api/system.windows.input.mouseaction?view=windowsdesktop-8.0) is more sophisticated
> 
>     2. the PointerGesture naming is rather confusing, since we have gesture recognition elsewhere. Maybe we should have PointerAction and PointerActionGesture? Not quite sure about the naming, need to discuss the API design more.

1. Main difference that I see - they have double clicks in this enum. We have click count, but only in pressed event. I added click count to this class, but not sure what to do with formatting. Maybe in case count > 1 use smth like Ctrl+Shift+Right*2 ?

2. Renamed to PointerActionGesture. Initially I called it PointerGesture in accordance with existing KeyGesture, because mechanisms are +- similar.

> Also, how was the problem with Ctrl+Click being used for list item selection resolved?
I think there is no problem with it. Ctrl+Click will select clicked ListBoxItem and open context menu on it, just like right click does. On mac toggling is performed with Cmd, so no conflict here.

herman-kirshin committed:
Added PointerActionGesture.Parse and more distinguishable button name formatting

MrJul commented:
I'm closing this because there is no consensus on the new API.

As mentioned in https://github.com/AvaloniaUI/Avalonia/pull/15842#issuecomment-2167461572, the implementation should probably be a macOS-specific gesture recognizer instead.
